### PR TITLE
Issue #141  - Add codefix for empty statement rule (S1116)

### LINKED ISSUE #141: Add codefix for empty statement rule (S1116)

### PR DESCRIPTION
**Pull request by tvsonar**
Fixes #141

### CODE REVIEW

File: src/Tests/SonarLint.UnitTest/Rules/EmptyStatementTest.cs [hunk 1]
```diff
@@ -32,5 +32,16 @@ public void EmptyStatement()
         {
             Verifier.VerifyAnalyzer(@"TestCases\EmptyStatement.cs", new EmptyStatement());
         }
+
+        [TestMethod]
+        [TestCategory("CodeFix")]
+        public void EmptyStatement_CodeFix()
```
Comment (tvsonar) [inline src/Tests/SonarLint.UnitTest/Rules/EmptyStatementTest.cs]: @dbolkensteyn This one fails at the moment, It needs the modification from #184

File: src/SonarLint/Rules/RedundantParenthesesCodeFixProvider.cs [hunk 2]
```diff
@@ -43,7 +43,7 @@ public sealed override FixAllProvider GetFixAllProvider()
         {
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/RedundantParenthesesCodeFixProvider.cs]: @tvsonar looks like this change is no longer required by this PR

File: src/SonarLint/Rules/EmptyStatementFixAllProvider.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                        async ct => fixAllContext.Document.WithSyntaxRoot(
+                            await GetFixedDocumentAsync(fixAllContext, fixAllContext.Document).ConfigureAwait(false))));
+                case FixAllScope.Project:
+                    return Task.FromResult(CodeAction.Create(EmptyStatementCodeFixProvider.Title,
+                        ct => GetFixedProjectAsync(fixAllContext, fixAllContext.Project)));
+                case FixAllScope.Solution:
+                    return Task.FromResult(CodeAction.Create(EmptyStatementCodeFixProvider.Title,
+                        ct => GetFixedSolutionAsync(fixAllContext)));
+            }
+            return null;
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/EmptyStatementFixAllProvider.cs]: @tvsonar it looks like there is one more scope `Custom` - Roslyn's fixed is throwing in all other cases. Perhaps we should just throw here? https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs#L214

File: src/SonarLint/Rules/EmptyStatementCodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    Title,
+                    c =>
+                    {
+                        var newRoot = RemoveUnusedCode(root, syntaxNode);
+                        return Task.FromResult(context.Document.WithSyntaxRoot(newRoot));
+                    },
+                    Title),
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/EmptyStatementCodeFixProvider.cs]: @tvsonar as discussed - let's remove this

File: src/SonarLint/Rules/EmptyStatementFixAllProvider.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            foreach (var newDoc in newDocuments)
+            {
+                solution = solution.WithDocumentSyntaxRoot(newDoc.Key, newDoc.Value.Result);
+            }
+            return solution;
+        }
+
+        private async static Task<SyntaxNode> GetFixedDocumentAsync(FixAllContext fixAllContext, Document document)
+        {
+            var diagnostics = await fixAllContext.GetDocumentDiagnosticsAsync(document).ConfigureAwait(false);
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/EmptyStatementFixAllProvider.cs]: `GetDocumentDiagnosticsAsync()` is filtered by `fixAllContext.DiagnosticsIds` which is perfect. Care must be taken when generalizing this fixall provider though.

**Comment by dbolkensteyn** [comment]
LGTM - but let's wait a bit before merging this into the master

**Pull request merged**